### PR TITLE
Array jesse

### PR DESCRIPTION
**Pull request by EsotericSquishyy**
Passing all tests

### CODE REVIEW

File: src/calc.cpp [hunk 1]
```diff
@@ -39,6 +39,63 @@ int main(){
                     std::cout << "false" << std::endl;
                 }
             }
+            else if (treeVal.type == ReturnType::ARRAY) {
+                auto array = std::get<std::shared_ptr<variableVal::Array>>(treeVal.value);
+                /*std::cout << "[";
```
Comment (EsotericSquishyy) [inline src/calc.cpp]: no need for this comment any more

File: src/lib/AST.cpp [hunk 2]
```diff
@@ -18,6 +18,42 @@ bool variableVal::operator == (const variableVal& rVal) const{
             return (std::get<bool>(value) == std::get<bool>(rVal.value));
         case ReturnType::FUNC:
             return (std::get<std::shared_ptr<Func>>(value)->mForest == std::get<std::shared_ptr<Func>>(rVal.value)->mForest);
+        case ReturnType::ARRAY: {
+            // Check if one variableVal is NOT an array
```
Comment (EsotericSquishyy) [inline src/lib/AST.cpp]: we already know both variable are the same type. I think we can do something like we did on line 20 but with arrays

File: src/lib/AST.cpp [hunk 3]
```diff
@@ -43,6 +79,10 @@ variableVal::Func::~Func(){
     mForest.reset();
 }
 
+/*variableVal::Array::~Array(){
+    // implement?
```
Comment (EsotericSquishyy) [inline src/lib/AST.cpp]: write a destructor like the Func destructor

File: src/lib/AST.cpp [hunk 4]
```diff
@@ -726,3 +905,123 @@ void TreeStatement::printInfix(int depth) const{
...
+
+// Returns shared_ptr< Array>, checks map to see if this ID is assigned to an array
+std::shared_ptr<variableVal::Array> TreeArray::getArray(std::unordered_map<std::string, variableVal>& vars){
+    std::vector<variableVal> tempArray;
+    for (auto& e : arrayElements) {
+        tempArray.push_back(e->evaluate(vars));
+    }
+    return std::make_shared<variableVal::Array>(tempArray);
+}
+
```
Comment (EsotericSquishyy) [inline src/lib/AST.cpp:944]: 10 newlines here

File: src/lib/AST.h [hunk 5]
```diff
@@ -51,8 +53,21 @@ struct variableVal {
...
+    struct Array {
+        /*
+        A struct to properly store all information pretaining to arrays
+        so a arrays can be assigned variables, reassigned, and properly handeled.
+        */
+
+        std::vector<variableVal> elements;
+        
+        Array(const std::vector<variableVal>& e) : elements(e) {}
+        //~Array();
```
Comment (EsotericSquishyy) [inline src/lib/AST.h]: make sure to uncomment if you write a destructor

**Review by EsotericSquishyy — COMMENTED** [review]
Looks good, a few formatting comments and a couple simplification comments

File: src/lib/AST.cpp [hunk 6]
```diff
@@ -419,6 +519,66 @@ variableVal TreeCall::evaluate(std::unordered_map<std::string, variableVal>& var
...
+        throw std::runtime_error("Runtime error: underflow."); 
+    }
+
+    variableVal popVal = array->elements[array->elements.size() - 1];
+    array->elements.pop_back();
+
+    return popVal;
+}
+
+
```
Comment (chloeandersen-ucsb) [inline src/lib/AST.cpp:584]: These helper functions look good!

**Comment by chloeandersen-ucsb** [comment]
Changes made! Documentation added

**Pull request merged**